README.md
This package implements the Timecode SMPTE spec.

## Features

A `Timecode` object has a framerate and a number of frames. Various mathematical operations
can be done on the object, and it will be able to understand how to convert its frame number
to a human readable SMPTE string like `01:23:45:01` where the form is `HH:MM:SS:FF` (hours,
minutes, seconds, frames).

`Timecode` objects can also be configured to use milliseconds instead of frames or to output
SMPTE code with fractional seconds instead of frame numbers. In either case, the output will
look like this `01:23:45.123` where the last three digits are milliseconds.

## Usage

Add to your `pubspec.yaml`.

```bash
$ dart pub add timecode
```

Import in your file and create a `Timecode` object.

```dart
import 'package:timecode/timecode.dart';

var timecode = Timecode(framerate: TimecodeFramerate(24));
print(timecode);
for (var i = 0; i < 100; i++) {
	timecode.next();
	print(timecode);
}
```

## Additional information

The class is well-documented, so you should be able to understand it easily by looking at the API documentation
or by reading the source code directly.

### Reference material

This work is based on the work of the following:

-   https://www.connect.ecuad.ca/~mrose/pdf_documents/timecode.pdf
-   https://www.davidheidelberger.com/2010/06/10/drop-frame-timecode/
-   https://pypi.org/project/pytimecode.py/


CHANGELOG.md
## 0.4.0

-   The Timecode object can now 'run' meaning it has functions to start, stop, pause, and resume an internal timer that will not only update the frame count according to real time, but will also output a stream of Timecode updates.

## 0.3.0

-   minor documentation improvements and a spelling correction.

## 0.2.0

-   Directly implemented the algorithms from https://www.davidheidelberger.com/2010/06/10/drop-frame-timecode/ but made a few modifications
-   Timecodes now handle drop frames correctly.

## 0.1.0

-   Initial Version... still needs some testing.


test/timecode_test.dart
import 'package:timecode/timecode.dart';
import 'package:test/test.dart';

void main() {
  group('Testing...', () {
    setUp(() {
      // Additional setup goes here.
    });

    test('Millis', () {
      final t1000 = Timecode(framerate: TimecodeFramerate(1000));
      expect(t1000.toString(), equals('00:00:00.000'));

      final t60 = Timecode(framerate: TimecodeFramerate(60));
      t60.forceFractionalSeconds = true;
      expect(t60.toString(), equals('00:00:00.000'));
    });

    test('Drop Frame Flags', () {
      final t2997 = Timecode(framerate: TimecodeFramerate(29.97));
      final t5994 = Timecode(framerate: TimecodeFramerate(59.94));
      for (var testable in [t2997, t5994]) {
        expect(testable.toString(), equals('00:00:00;00'));
        expect(testable.isDropFrame, isTrue);
      }
    });

    test('Non Drop Frame Flags', () {
      final t24 = Timecode(framerate: TimecodeFramerate(24));
      final t2398 = Timecode(framerate: TimecodeFramerate(23.98));
      final t30 = Timecode(framerate: TimecodeFramerate(30));
      final t60 = Timecode(framerate: TimecodeFramerate(60));
      final t2997 = Timecode(framerate: TimecodeFramerate(29.97, forceNonDropFrame: true));
      final t5994 = Timecode(framerate: TimecodeFramerate(59.94, forceNonDropFrame: true));

      for (var testable in [t24, t2398, t30, t60, t2997, t5994]) {
        expect(testable.toString(), equals('00:00:00:00'));
        expect(testable.isDropFrame, isFalse);
      }
    });

    test('Constructors', () {
      var t = Timecode.atTimecode('00:00:59:00');
      expect(t.toString(), equals('00:00:59:00'));
      expect(t.framerate.fps, equals(24));
      expect(t.frameCount, equals(24 * 59));

      t = Timecode.atTimecode('00:00:59;00');
      expect(t.toString(), equals('00:00:59;00'));
      expect(t.framerate.fps, equals(29.97));
      expect(t.frameCount, equals((30 * 59).floor()));

      t = Timecode.atTimecode('00:00:59.000');
      expect(t.toString(), equals('00:00:59.000'));
      expect(t.framerate.fps, equals(1000));
      expect(t.frameCount, equals(1000 * 59));

      t = Timecode.atSeconds(10, framerate: TimecodeFramerate(24));
      expect(t.toString(), equals('00:00:10:00'));

      t = Timecode.atSeconds(10, framerate: TimecodeFramerate(29.97));
      expect(t.toString(), equals('00:00:09;29'));

      t = Timecode();
      expect(t.toString(), equals('00:00:00:00'));
      expect(t.frameCount, equals(0));
    });

    test('Drop Frame Increments with 29.97', () {
      var t = Timecode.atTimecode('00:00:59;00', framerate: TimecodeFramerate(29.97));
      t.addFrames(29);
      expect(t.toString(), equals('00:00:59;29'));
      t.next();
      expect(t.toString(), equals('00:01:00;02'));
      t.next();
      expect(t.toString(), equals('00:01:00;03'));
    });

    test('Forced Non Drop Frame Increments with 29.97', () {
      var t = Timecode.atTimecode('00:00:59:00', framerate: TimecodeFramerate(29.97, forceNonDropFrame: true));
      t.addFrames(29);
      expect(t.toString(), equals('00:00:59:29'));
      t.next();
      expect(t.toString(), equals('00:01:00:00'));
      t.next();
      expect(t.toString(), equals('00:01:00:01'));
    });

    test('Non Drop Frame Increments', () {
      var t = Timecode.atSeconds(10, framerate: TimecodeFramerate(24));
      t.addFrames(23);
      expect(t.toString(), equals('00:00:10:23'));
      t.next();
      expect(t.toString(), equals('00:00:11:00'));
    });
  });
}


example/timecode_example.dart
import 'package:timecode/timecode.dart';

void main() {
  // var fps = TimecodeFramerate(29.97);
  // var testStrings = {
  //   '00:00:59;29': 1799,
  //   '00:01:00;02': 1800 * 1,
  //   '00:02:00;04': 1800 * 2,
  //   '00:03:00;06': 1800 * 3,
  //   '00:04:00;08': 1800 * 4,
  //   '00:05:00;10': 1800 * 5,
  //   '00:06:00;12': 1800 * 6,
  //   '00:07:00;14': 1800 * 7,
  //   '00:08:00;16': 1800 * 8,
  //   '00:09:00;18': 1800 * 9,
  //   '00:10:00;00': 17982,
  //   '01:10:00;00': 125874,
  // };
  // for (var tc in testStrings.keys) {
  //   var frame = testStrings[tc]!;
  //   var t1 = Timecode.atTimecode(tc, framerate: fps);
  //   print('$tc - $t1 - ${t1.frameCount} - $frame');
  //   var t2 = Timecode(framerate: fps, startFrames: frame);
  //   print('$tc - $t1 - ${t1.frameCount} - $frame');
  //   print(t1 == t2);
  // }

  var timecode = Timecode.atTimecode('00:00:59:00', framerate: TimecodeFramerate(29.97, forceNonDropFrame: true));
  print('$timecode - ${timecode.frameCount}');
  for (var i = 0; i < 10; i++) {
    timecode.next();
    print('$timecode - ${timecode.frameCount}');
  }

  timecode = Timecode.atTimecode('00:01:00;02', framerate: TimecodeFramerate(29.97, forceNonDropFrame: false));
  print('$timecode - ${timecode.frameCount}');
  for (var i = 0; i < 10; i++) {
    timecode.next();
    print('$timecode - ${timecode.frameCount}');
  }

  // print(Timecode.atSeconds(0, framerate: fps)); // -> 00:00:00;00
  // print(Timecode.atSeconds(1, framerate: fps)); // -> 00:00:00;29
  // print(Timecode.atSeconds(10, framerate: fps)); // -> 00:00:09;29
}


lib/src/timecode_base.dart
import 'dart:async';

const double defaultFPS = 24;

/// Some parts have been copied / inspired by the pytimecode library by Joshua Banton
///
/// https://pypi.org/project/pytimecode.py/
///
/// Notes: *There is a 24 hour SMPTE Timecode limit, so if your time exceeds that limit, it will roll over.
///
/// [Timecode] Does all the calculation over frames, so the main data it holds is
/// frames, then when required it converts the frames to a timecode by
/// using the frame rate setting.
///
/// Note: Drop frame timecodes operate by reporting higher frame numbers
/// than really exist so that 29.97 timecodes stay in sync with 30 fps ones, etc.
///
/// Drop Frame Example:
///
/// at 29.97 fps, the next frame after 00:00:59;29 will be 00:01:00;02 (skipping 0 & 1)
/// See the document here: https://www.connect.ecuad.ca/~mrose/pdf_documents/timecode.pdf
///
/// Drop Frame Calculations are done by implementing code found here:
/// - https://www.davidheidelberger.com/2010/06/10/drop-frame-timecode/
/// - https://robwomack.com/timecode-calculator/
class Timecode extends Object {
  TimecodeFramerate framerate;

  // convenience getters ---
  bool get isDropFrame => framerate.isDropFrame;
  int get integerFps => framerate.integerFps;
  double get fps => framerate.fps;

  // display frames as fractions of a second instead of frame numbers
  bool get isMillis => integerFps == 1000;
  bool forceFractionalSeconds = false;

  /// keeps track of the actual frame count
  int get frameCount => _frameCount;
  set frameCount(int n) {
    if (n < 0) throw TimecodeException('frames must be given in integers greater than zero. Received: $n');
    _frameCount = n;
    _recomputeParts();
    _streamController.add(true);
  }

  int _frameCount = 0;

  /// this is the frame to use whenever we `reset` this timecode
  int startingFrame = 0;

  /// reports whether the timecode has a start time
  bool get started => startTime != null;

  /// the real time when this timecode was "started"
  DateTime? startTime;

  /// reports whether the timecode internal timer is running
  bool get running => _timer?.isActive ?? false;

  Timer? _timer;
  final StreamController<bool> _streamController = StreamController.broadcast();

  /// a stream of updates when the timecode internal timer is running
  /// NOTE: The stream gets another `true` value every time it updates.
  Stream<bool> get updateStream => _streamController.stream;

  /// the real timecode parts get updated whenever frames are changed
  TimecodeData parts = TimecodeData();

  // these are the variables
  int get hh => parts.hh;
  int get mm => parts.mm;
  int get ss => parts.ss;

  /// will always honor the drop frame setting in the framerate
  int get ff => parts.ff;

  /// fractional seconds in milliseconds honors drop frame
  int get frac => parts.frac;

  /// millis reflects the realtime duration of this instance
  int get millis => parts.millis;

  // helper for `toString` method
  String get frameDelimiter => isDropFrame
      ? ';'
      : isMillis || forceFractionalSeconds
          ? '.'
          : ':';

  /// Creates a new Timecode instance.
  ///
  /// @parameters:
  ///
  /// [framerate] a [TimecodeFramerate] instance.
  ///
  /// [startFrames] : Optional starting offset in frames. Use this to force the
  /// timecode to start at a specific frame value (in real frames).
  Timecode({
    TimecodeFramerate? framerate,
    int startFrames = 0,
  }) : framerate = framerate ?? TimecodeFramerate(defaultFPS) {
    startingFrame = _frameCount = startFrames;
    _recomputeParts();
  }

  /// Creates a timecode with a specific timecode string and framerate.
  ///
  /// If [framerate] is omitted, the following defaults will be used:
  /// - default timecodes (00:00:00:00) -> 24 fps
  /// - drop frame timecodes (00:00:00;00) -> 29.97 fps
  /// - ms timecodes (00:00:00.000) -> 1000 fps
  ///
  /// NOTE: When using drop frame framerates, some timecode strings are considered
  /// invalid and will be corrected to valid timecodes. Consider the following examples.
  ///
  /// ```dart
  /// var fps = TimecodeFramerate(29.97);
  /// Timecode.atTimecode('00:00:59;29', framerate: fps); // -> 00:00:59;29
  /// Timecode.atTimecode('00:01:00;00', framerate: fps); // -> 00:01:00;02 (!!)
  /// Timecode.atTimecode('00:01:00;01', framerate: fps); // -> 00:01:00;02 (!!)
  /// Timecode.atTimecode('00:01:00;02', framerate: fps); // -> 00:01:00;02
  /// ```
  factory Timecode.atTimecode(String timecodeString, {TimecodeFramerate? framerate}) {
    double def = 24;
    if (timecodeString.contains(';')) {
      def = 29.97;
    } else if (timecodeString.contains('.')) {
      def = 1000;
    }
    framerate ??= TimecodeFramerate(def);
    var startFrames = parseToFrames(timecodeString, framerate: framerate);
    return Timecode(framerate: framerate, startFrames: startFrames);
  }

  /// Creates a timecode at a specific number of seconds with a specified framerate.
  ///
  /// NOTE: When using fractional framerates, the resulting timecode may not match
  /// what you expect. Consider the following examples.
  ///
  /// ```dart
  /// var fps = TimecodeFramerate(29.97);
  /// Timecode.atSeconds(0, framerate: fps); // -> 00:00:00;00
  /// Timecode.atSeconds(1, framerate: fps); // -> 00:00:00;29
  /// Timecode.atSeconds(10, framerate: fps); // -> 00:00:09;29
  /// ```
  factory Timecode.atSeconds(double seconds, {TimecodeFramerate? framerate}) {
    framerate ??= TimecodeFramerate(defaultFPS);
    var startFrames = framerate.realSecondsToFrames(seconds);
    return Timecode(framerate: framerate, startFrames: startFrames);
  }

  /// The default format for a [timecodeString] is `HH:MM:SS:FF` where
  /// `FF` stands for the number of frames after the current second and
  /// may be `0 <= FF < fps`.
  ///
  /// When a drop frame encoding has been employed, the format will be
  /// `HH:MM:SS;FF` (final semicolon) or `HH;MM;SS;FF` (all semicolons).
  ///
  /// An alternate form looks like this `HH:MM:SS.SSS` where the seconds
  /// are specified in floating point form. This usually indicates the
  /// fps is 1000 (millisecond) but can be used with any framerate.
  static int parseToFrames(String timecodeString, {TimecodeFramerate? framerate}) {
    framerate ??= TimecodeFramerate(defaultFPS);
    var isDropFrame = timecodeString.contains(';');
    if (isDropFrame != framerate.isDropFrame) {
      throw TimecodeException(
          'timecode string and framerate mismatch. Framerates that employ drop frame encoding (29.97 and 59.94) use a timecode format like this 00:00:00;00');
    }
    var parts = timecodeString.replaceAll(';', ':').split(':').map<double>((e) => double.tryParse(e) ?? 0).toList();

    // if the timecode were fractional, there will be only three elements
    if (parts.length < 3) return 0;

    var hh = parts[0].toInt();
    var mm = parts[1].toInt();
    var ss = parts[2];
    int ff = (parts.length == 4) ? parts[3].toInt() : 0;

    var totalMinutes = mm + hh * 60;
    var totalSeconds = ss + totalMinutes * 60;
    var timecodeFrames = ff + framerate.timecodeSecondsToFrames(totalSeconds);
    return framerate.realFrames(timecodeFrames);
  }

  @override
  operator ==(Object other) {
    if (other is Timecode) {
      return other.millis == millis;
    }
    return false;
  }

  operator >(Timecode other) {
    return other.millis > millis;
  }

  operator <(Timecode other) {
    return other.millis < millis;
  }

  Timecode operator +(Timecode other) {
    var m = millis + other.millis;
    var seconds = m / 1000;
    return Timecode.atSeconds(seconds, framerate: framerate);
  }

  Timecode operator -(Timecode other) {
    var m = millis - other.millis;
    var seconds = m / 1000;
    return Timecode.atSeconds(seconds, framerate: framerate);
  }

  Timecode operator *(int multiplier) {
    var m = millis * multiplier;
    var seconds = m / 1000;
    return Timecode.atSeconds(seconds, framerate: framerate);
  }

  Timecode operator /(int divisor) {
    var m = millis / divisor;
    var seconds = m / 1000;
    return Timecode.atSeconds(seconds, framerate: framerate);
  }

  @override
  int get hashCode => _frameCount.hashCode;

  void _recomputeParts() {
    parts = framerate.parseFrames(frameCount);
  }

  /// can use fractional seconds
  void addSeconds(double seconds) => frameCount = frameCount + framerate.timecodeSecondsToFrames(seconds);
  void addFrames(int n) => frameCount = frameCount + n;
  void subFrames(int n) => frameCount = frameCount - n;
  void multFrames(int n) => frameCount = frameCount * n;
  void divFrames(int n) => frameCount = frameCount ~/ n;
  void next() => addFrames(1);
  void back() => frameCount > 0 ? subFrames(1) : null;

  @override
  String toString() {
    var hhs = hh.toString().padLeft(2, '0');
    var mms = mm.toString().padLeft(2, '0');
    var sss = ss.toString().padLeft(2, '0');
    var ffs = (forceFractionalSeconds || isMillis) ? frac.toString().padLeft(3, '0') : ff.toString().padLeft(2, '0');
    var delim = frameDelimiter;
    return '$hhs:$mms:$sss$delim$ffs';
  }

  /// Will "start" a timecode by recording the current start time.
  /// Subsequent calls to `update` will then update the frames
  /// from the elapsed time since this start time.
  void start({bool fromCurrentFrame = false}) {
    stop(andReset: !fromCurrentFrame);
    startTime = DateTime.now();
    resume();
  }

  /// Will reset the timecode instance to the frame identified by
  /// `startingFrame` which was set when the object is created
  /// but may have been updated since then.
  void reset() {
    frameCount = startingFrame;
  }

  /// Will stop a timecode and reset the `startTime` to `null`
  void stop({bool andReset = true}) {
    _timer?.cancel();
    startTime = null;
    if (andReset) reset();
  }

  /// Will pause the internal timecode timer so that it can be resumed.
  void pause() {
    _timer?.cancel();
  }

  /// Will resume a paused internal timer or will start one now.
  void resume() {
    if (!started) {
      start();
      return;
    }
    _timer?.cancel();
    var tenthOfAFrame = 0.1 * 1000000 / fps;
    _timer = Timer.periodic(
      Duration(microseconds: tenthOfAFrame.toInt()),
      (timer) {
        update();
      },
    );
  }

  /// Will be called by the internal timer, or may be called at any time.
  void update() {
    if (startTime == null) throw Exception('Must call `start()` before calling this.');
    var now = DateTime.now();
    var diff = now.difference(startTime!).inMilliseconds;
    if (diff < 0) throw Exception('Back in Time: Current time is before the start time!');
    frameCount = framerate.realSecondsToFrames(diff / 1000);
  }

  /// Will run an update and then return the string representation of this timecode.
  String now() {
    update();
    return toString();
  }
}

/// The framerate of a Timecode instance. It may be based on any
/// arbitrary floating point value, but usually will be one of the industry
/// standard framerates: 23.976, 23.98, 24, 25, 29.97, 30, 50, 59.94, 60, 1000.
/// If the framerate is either 29.97 or 59.94, the internal [isDropFrame] flag
/// will be set unless `forceNonDropFrame` is specified in the constructor.
///
/// This class handles all calculations regarding framerates. It handles
/// the drop frame calculations, and can convert between seconds and frames.
///
/// Drop Frame Calculations are done by implementing code found here:
/// - https://www.davidheidelberger.com/2010/06/10/drop-frame-timecode/
/// - https://robwomack.com/timecode-calculator/
///
/// However, Rob Womack's implementation uses the integer fps when
/// it should have used the full fps. David Heidelberger's code is likewise
/// ambiguous on when to use integer fps and when to use the full fps.
class TimecodeFramerate {
  final double fps;
  final int integerFps;
  final bool isDropFrame;

  /// [skippedFrameNumbersPerMinute] is the number of frames to drop every minute
  /// and is computed as the nearest integer to 6% of the full fps
  late final int skippedFrameNumbersPerMinute;
  late final int realFramesPerHour;
  late final int realFramesPer24Hours;
  late final int realFramesPer10Minutes;
  late final int timecodeFramesPerMinute;

  TimecodeFramerate._(this.fps, this.integerFps, this.isDropFrame) {
    // these are calculations that are used in dropFrame conversions
    skippedFrameNumbersPerMinute = isDropFrame ? (fps * 0.0666666).round() : 0;
    realFramesPerHour = (fps * 60 * 60).round();
    realFramesPer24Hours = realFramesPerHour * 24;
    realFramesPer10Minutes = (fps * 60 * 10).round();

    // timecodeFramesPerMinute uses the integerFps because it also accounts for dropFrames
    timecodeFramesPerMinute = (integerFps * 60) - skippedFrameNumbersPerMinute;
  }

  /// Will create a [TimecodeFramerate] instance. [fps] may be any
  /// arbitrary floating point value, but usually will be one of the industry
  /// standard framerates: 23.976, 23.98, 24, 25, 29.97, 30, 50, 59.94, 60, 1000.
  /// If the framerate is either 29.97 or 59.94, the internal [isDropFrame] flag
  /// will be set unless [forceNonDropFrame] is specified in the constructor.
  factory TimecodeFramerate(double fps, {forceNonDropFrame = false}) {
    var integerFramerate = fps.round();
    var isDropFrame = false;

    // compute the integer framerate and drop frame status if needed.
    switch ((fps * 100).round()) {
      case 2997:
        isDropFrame = !forceNonDropFrame;
        break;
      case 5994:
        isDropFrame = !forceNonDropFrame;
        break;
    }
    return TimecodeFramerate._(
      fps,
      integerFramerate,
      isDropFrame,
    );
  }

  double rollover(double seconds) {
    while (seconds < 0) {
      seconds += 60 * 60 * 24;
    }
    return seconds % (60 * 60 * 24);
  }

  /// returns real frame count from seconds of real time
  /// and does not account for drop frames
  /// if seconds > 24 hours, rolls over
  int realSecondsToFrames(double seconds) {
    seconds = rollover(seconds);
    return (seconds * fps).floor();
  }

  /// returns timecode frame count from timecode seconds.
  /// Timecode seconds are always based on the integerFps
  /// and that's why drop frames are needed
  /// will roll over after 24 hours
  int timecodeSecondsToFrames(double seconds, {ignoreDropFrame = false}) {
    seconds = rollover(seconds);
    return seconds.round() * integerFps;
    // return (seconds * ((isDropFrame && !ignoreDropFrame) ? integerFps : fps)).floor();
  }

  /// Code modified for Dart and split into two functions
  /// from https://www.davidheidelberger.com/2010/06/10/drop-frame-timecode/
  /// Also consider the implementation here:
  /// http://www.andrewduncan.net/timecodes/
  TimecodeData parseFrames(int frameCount, {ignoreDropFrame = false}) {
    frameCount = timecodeFrames(frameCount, ignoreDropFrame: ignoreDropFrame);

    int frRound = fps.round();
    int ff = frameCount % frRound;
    int ss = (frameCount ~/ frRound) % 60;
    int mm = ((frameCount ~/ frRound) ~/ 60) % 60;
    int hh = (((frameCount ~/ frRound) ~/ 60) ~/ 60);
    int millis = (1000 * frameCount / frRound).floor(); // total time
    int frac = millis % 1000;

    // this is the "corrected" frame data
    return TimecodeData(hh, mm, ss, ff, frac, millis);
  }

  /// timecode frames will be equal to or larger than real frames.
  /// Example: in 29.97 fps, after each minute, two frames will be added to
  /// the timecode frame count, except after 10 minutes, no frames are added
  /// This function adds in the extra timecode frames.
  ///
  /// from https://www.davidheidelberger.com/2010/06/10/drop-frame-timecode/
  int timecodeFrames(int frameCount, {ignoreDropFrame = false}) {
    // always handle frameCount rollover

    // do not accept negative frame numbers
    while (frameCount < 0) {
      frameCount = realFramesPer24Hours + frameCount;
    }
    // greater than 24 hours rolls over to 0
    frameCount = frameCount % realFramesPer24Hours;

    if (isDropFrame && !ignoreDropFrame) {
      var dropFrames = skippedFrameNumbersPerMinute;
      int d = frameCount ~/ realFramesPer10Minutes;
      int m = frameCount % realFramesPer10Minutes;

      if (m > dropFrames) {
        frameCount = frameCount + (dropFrames * 9 * d) + dropFrames * ((m - dropFrames) ~/ timecodeFramesPerMinute);
      } else {
        frameCount = frameCount + dropFrames * 9 * d;
      }
    }

    return frameCount;
  }

  /// timecode frames will be equal to or larger than real frames.
  /// Example: in 29.97 fps, after each minute, two frames will be added to
  /// the timecode frame count, except after 10 minutes, no frames are added
  /// This code removes the extra timecode frames to get the real underlying
  /// frame count.
  int realFrames(int timecodeFrames) {
    var frames = timecodeFrames;

    if (isDropFrame) {
      var seconds = frames ~/ integerFps;
      var timecodeMinutes = seconds ~/ 60;
      var minutesWithDroppedFrames = timecodeMinutes - (timecodeMinutes ~/ 10);

      // Remove the extra timecode frames (the `drop frames`)
      frames -= minutesWithDroppedFrames * skippedFrameNumbersPerMinute;
    }
    return frames;
  }
}

/// Timecode data that is ignorant of framerate.
/// We use this to pass around structured timecode data
class TimecodeData {
  // these are the variables
  final int _hh;
  int get hh => _hh;
  final int _mm;
  int get mm => _mm;
  final int _ss;
  int get ss => _ss;
  final int _ff;
  int get ff => _ff;
  final int _frac; // fractional seconds in milliseconds
  int get frac => _frac;
  final int _millis;
  int get millis => _millis;

  const TimecodeData([
    this._hh = 0,
    this._mm = 0,
    this._ss = 0,
    this._ff = 0,
    this._frac = 0,
    this._millis = 0,
  ]);
}

class TimecodeException implements Exception {
  String message;
  TimecodeException(this.message);
  @override
  String toString() {
    return "TimecodeException: $message";
  }
}
